src/api/endpoint/api/v1/token.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * @prettier
+ */
+
+// @flow
+
+import { HTTP_STATUS, SUPPORT_ERRORS } from '../../../../lib/constants';
+import { ErrorCode } from '../../../../lib/utils';
+import { getApiToken } from '../../../../lib/auth-utils';
+import { stringToMD5 } from '../../../../lib/crypto-utils';
+import logger from '../../../../lib/logger';
+
+import type { $Response, Router } from 'express';
+import type { $NextFunctionVer, $RequestExtend, IAuth } from '../../../../../types';
+
+// https://github.com/npm/npm-profile/blob/latest/lib/index.js
+export default function(route: Router, auth: IAuth, storage: IStorageHandler, config: Config) {
+  route.get('/-/npm/v1/tokens', function(req: $RequestExtend, res: $Response, next: $NextFunctionVer) {
+    /**
+     * table.push([
+        token.id,
+        token.token + '…',
+        String(token.created).slice(0, 10),
+        token.readonly ? 'yes' : 'no',
+        token.cidr_whitelist ? token.cidr_whitelist.join(', ') : ''
+      ])
+     */
+    const fake = [
+      {
+        token: '1232f6',
+        key: 'key2',
+        cidr_whitelist: null,
+        readonly: false,
+        created: '2018-10-05T21:49:39.457Z',
+        updated: '2018-10-05T21:49:39.457Z',
+      },
+      {
+        token: '12b000',
+        key: 'key2',
+        cidr_whitelist: null,
+        readonly: false,
+        created: '2018-08-04T09:04:17.194Z',
+        updated: '2018-08-04T09:04:17.194Z',
+      },
+    ];
+
+    res.status(HTTP_STATUS.OK);
+    // return next({ message: API_ERROR.MUST_BE_LOGGED });
+    next({
+      objects: fake,
+      urls: {
+        next: '', // TODO pagination
+      },
+    });
+  });
+
+  route.post('/-/npm/v1/tokens', function(req: $RequestExtend, res: $Response, next: $NextFunctionVer) {
+    const { password, readonly, cidr_whitelist } = req.body;
+    const { name } = req.remote_user;
+
+    if (typeof readonly !== 'boolean' || !Array.isArray(cidr_whitelist)) {
+      next(ErrorCode.getCode(HTTP_STATUS.BAD_DATA, SUPPORT_ERRORS.PARAMETERS_NOT_VALID));
+      return;
+    }
+
+    auth.authenticate(name, password, async (err, user: RemoteUser) => {
+      if (err) {
+        const errorCode = err.message ? HTTP_STATUS.UNAUTHORIZED : HTTP_STATUS.INTERNAL_ERROR;
+        next(ErrorCode.getCode(errorCode, err.message));
+      } else {
+        req.remote_user = user;
+
+        if (typeof storage.saveToken !== 'function') {
+          next(ErrorCode.getCode(HTTP_STATUS.NOT_IMPLEMENTED, SUPPORT_ERRORS.STORAGE_NOT_IMPLEMENT));
+          return;
+        }
+
+        try {
+          const token = await getApiToken(auth, config, user, password);
+          const key = stringToMD5(token);
+
+          const saveToken = {
+            user: name,
+            viewToken: `${token.substr(0, 3)}...${token.substr(-3)}`,
```

Comment:
`${token.substr(0, 3)}...${token.substr(-3)}` this should be in a utility thus we can test it easily. We had to add unit test at the end of all of this.